feat(parser): perform case sensitive parsing

mgechev opened the pull request:
Monkey patches the parse5 tokenizer in order to allow case sensitive parsing. This way we can use case sensitive selector in order to strip content that should not belong to the App Shell.

jeffbcross commented on app-shell/src/app/shell-parser/template-parser/parse5-template-parser.spec.ts:
```diff
@@ -0,0 +1,37 @@
+<html>
+<body>
+  <Div>
+    <Section Title="Bar"></Section>
```
Can you add more variety of mixed casing to the template?

mgechev replied:
Done!

jeffbcross commented on app-shell/src/app/shell-parser/template-parser/parse5-template-parser.ts:
```diff
@@ -3,6 +3,48 @@ import {TemplateParser} from './template-parser';
+// won't have access to the tokenizer in order to patch
+// it runtime. Because of that we distribute the entire
+// Runtime Parser as a single bundle which includes parse5.
+Tokenizer.prototype.getNextToken = function () {
```
I'm having trouble reading through this function (probably because it's end of day on a Friday :)). Can you add a few unit tests for this?

mgechev replied:
Yes, here are [a few tests](https://github.com/angular/mobile-toolkit/pull/68/files#diff-385ca2e7c2233573bbf0b0e65ce27dc6R1).

jeffbcross commented:
A couple of comments added!

By design, Angular doesn't preserve the `shellRender` and `shellNoRender`  selector attributes on the compiled elements. Refresh my memory, is the assumption of this PR that at some point in the near future we would modify the directives to actually add `shellRender` attributes on the nodes during pre-render time?

mgechev commented:
@jeffbcross I addressed the comments you had. As we discussed, I'll update the `shellRender` and `shellNoRender` directives, and their specs, in order to add the corresponding attributes.

jeffbcross commented:
Opened a couple of follow-up issues: #71, #70

jeffbcross commented on app-shell/package.json:
```diff
@@ -10,7 +10,9 @@
     "e2e": "protractor",
-    "build_publish": "rm -rf dist && tsc -p src/tsconfig.publish.es5.json && tsc -p src/tsconfig.publish.es6.json && cp src/package.json dist/package.json"
+    "clean": "rm -rf dist",
+    "build": "ng build && tsc -p src/tsconfig.publish.es5.json && tsc -p src/tsconfig.publish.es6.json && cp src/package.json dist/app/package.json && browserify dist/app/shell-parser/index.js -s shellParserFactory > dist/app/shell-parser.js && rm -rf dist/app/shell-parser && rm -rf dist/app/vendor",
```
Needs browserify dependency installed. I can take care of this.

mgechev replied:
Done!

jeffbcross commented on app-shell/src/app/shell-parser/template-parser/parse5/tokenizer-patch.ts:
```diff
@@ -0,0 +1,42 @@
+      let val = token[prop];
+      let last = val[val.length - 1];
+      if (last !== char) {
+        token[prop] = val.substring(0, val.length - 1) + last.toUpperCase();
```
Tests get errors on this line if last is undefined (`TypeError: Cannot read property 'toUpperCase' of undefined`). If I change to `if (last && last !== char)` then tests pass. Safe change to make?

mgechev replied:
Yes, it is safe.

jeffbcross commented on app-shell/src/app/shell-parser/template-parser/parse5/tokenizer-patch.ts:
```diff
@@ -0,0 +1,42 @@
+      this[this.state](cp);
+    }
+    switch (this.state) {
+      case 'TAG_NAME_STATE':
```
Why do we need case sensitivity for tag names?

mgechev replied:
I did it in order to be consistent with the Angular parser.

jeffbcross commented on app-shell/src/app/shell-parser/template-parser/parse5/tokenizer-patch.spec.ts:
```diff
@@ -0,0 +1,53 @@
+var Tokenizer = require('../../../../vendor/parse5/lib/tokenizer');
+
+const caseSensitiveTemplate =
+`
```
Delete

jeffbcross commented on app-shell/src/app/shell-parser/template-parser/parse5/tokenizer-patch.spec.ts:
```diff
@@ -0,0 +1,53 @@
+    expect(closeDiv.tagName).toBe('DiV');
+  });
+
+  it('should preserve case sensitivity of complex eements', () => {
```
mis-spelling

jeffbcross commented:
LGTM. Once comments are addressed, plz merge.
